### PR TITLE
Add ComAwareEventInfo tests

### PR DESCRIPTION
**Pull request by hughbe**
Extracted from #21140

### CODE REVIEW

**Comment by hughbe** [comment]
@joperezr I was wondering if you could explain the failures here for netfx and UAP, sorry...

I'm using RefEmit in one of the tests which isn't supported in UAP, so do I need to add a configuration for this UAP in the test project? Also, I don't understand why netfx would fail to compile: 

> error CS0246: The type or namespace name 'AssemblyBuilder' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\uapaot_debug_---28c5216e\src\System.Runtime.InteropServices\tests\System.Runtime.InteropServices.Tests.csproj]

Thanks!

**Comment by danmosemsft** [comment]
@hughbe refemit is not part of netstandard yet your test Configuration.props lists netstandard amongst its configurations: that configuration will not resolve refemit types like AssemblyBuilder.
When you build uap/uapaot it will build the netstandard configuration if there is no uap/uapaot configuration.

**Comment by hughbe** [comment]
I see! Another question: currently the file is:
```xml
<BuildConfigurations>
      netstandard-Unix;
      netstandard-Windows_NT;
      netcoreapp-Unix;
      netcoreapp-Windows_NT;
</BuildConfigurations>
```

why do we need to duplicate the Unix/Windows_NT parts. Can this be consolidated to

```xml
<BuildConfigurations>
      netcoreapp;
      netfx;
      uap;
      uapaot;
</BuildConfigurations>
```

**Comment by danmosemsft** [comment]
I think this
```
<BuildConfigurations>
      netstandard-Unix;
      netstandard-Windows_NT;
      netcoreapp-Unix;
      netcoreapp-Windows_NT;
</BuildConfigurations>
```
means build for netstandard, and build for netcoreapp, each separately for windows and unix, producing 4 binaries. If you omitted the windows and Unix suffixes, then it would produce a single shared binary for each, as opposed to (presumably) distinct binaries to package individually

```
<BuildConfigurations>
      netcoreapp;
      netfx;
      uap;
      uapaot;
</BuildConfigurations>
```
means build 4 binaries, each platform-agnostic (although in practice the last three are windows only).

A uap or uapaot configuration is not the same as a netstandard configuration -- the latter compiles against netstandard.dll, so it can work anywhere ,the former two build against assemblies named like system.private.corelib, system.runtime etc. More importantly building for netstandard creates a single assembly that will be packaged for any platform. 

In an ideal world maybe every project would have exactly one configuration, netstandard. However in practice there may need to be platform-specific code, or there may be API used that isn't available in netstandard. In that case the configuration must be platform specific or target framework specific respectively. What we package and ship for uapaot use, for example, is partly netstandard configuration binaries, a bunch of uap configuration, and a few uapaot.

@weshaggard @joperezr did I misstate any of this?

File: src/System.Runtime.InteropServices/tests/Configurations.props [hunk 1]
```diff
@@ -2,8 +2,11 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <BuildConfigurations>
-      netstandard;
-      netcoreapp;
+      netcoreapp-Windows_NT;
+      netcoreapp-Unix;
```
Comment (danmosemsft) [inline src/System.Runtime.InteropServices/tests/Configurations.props]: @joperezr is this the right set of configs? The src\configurations.props has
```xml
    <BuildConfigurations>
      uapaot-Windows_NT;
      netcoreapp-Windows_NT;
      netcoreapp-Unix;
      uap-Windows_NT;
    </BuildConfigurations>
```

So I think this is right, but is there any value in the OS suffixes for the uap/uapaot ones?

Comment (weshaggard) [inline src/System.Runtime.InteropServices/tests/Configurations.props]: >So I think this is right, but is there any value in the OS suffixes for the uap/uapaot ones?

OS suffixes should be avoided unless there is a reason that we need to condition code for a given OS. From a quick look at the code in this PR I don't think we need OS suffixes for any of the configurations. I don't see any #ifdefs or conditioned source files based on Windows or Unix for netcoreapp so they shouldn't be needed.

Comment (joperezr) [inline src/System.Runtime.InteropServices/tests/Configurations.props]: Unfortunatelly there is one condition inside the csproj that does use `TargetsWindows` which means the suffixes are required on the cases where we actually want that condition to be true.

File: src/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.Tests.csproj [hunk 2]
```diff
@@ -6,37 +6,42 @@
...
-    <Compile Include="System\Runtime\InteropServices\MarshalTests.cs" />
-    <!-- Helpers -->
-    <Compile Include="$(CommonTestPath)\System\PlatformDetection.cs">
-      <Link>Common\System\PlatformDetection.cs</Link>
-    </Compile>
     <Compile Include="System\Runtime\InteropServices\AllowReversePInvokeCallsAttributeTests.cs" />
     <Compile Include="System\Runtime\InteropServices\CallingConventionTests.cs" />
     <Compile Include="System\Runtime\InteropServices\ComAliasNameAttributeTests.cs" />
+    <Compile Include="System\Runtime\InteropServices\ComAwareEventInfoTests.cs" />
+    <Compile Include="System\Runtime\InteropServices\ComAwareEventInfoTests.Windows.cs" Condition="'$(TargetGroup)' == 'netfx' OR ('$(TargetsWindows)' == 'true' AND '$(TargetGroup)' != 'uap' AND '$(TargetGroup)' != 'uapaot')" />
```
Comment (joperezr) [inline src/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.Tests.csproj]: If you don't count this condition, we only really need a netstandard config and a netcoreapp one (for the definition of the 'netcoreapp' constant). Also, the condition seems over complicated to me, when is it exactly that we want to run these tests? is it only for netcoreapp|Windows and netfx|Windows?

Comment (hughbe) [inline src/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.Tests.csproj]: > is it only for netcoreapp|Windows and netfx|Windows?

Exactly. This test uses COM, with is only supported on netfx and netcoreapp

Comment (joperezr) [inline src/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.Tests.csproj]: I see, in that case, This is the set of configurations that I would have for this project:
```xml
<BuildConfigurations>
      netstandard;
      netcoreapp-Windows_NT;
      netfx-Windows_NT;
</BuildConfigurations>
```
That way we would only be building three different binaries depending on where we are building, and our configuration system will know how to pick the right one, and we would keep only the configurations that we actually need. Note that we could simplify it even more if we decided it's ok to run these tests on uap|Windows and uapaot|Windows, cause we could reduce to only have two configurations: `netstandard; netstandard-Windows_NT;`

File: src/System.Runtime.InteropServices/tests/Configurations.props (hunk 1 above)
Comment (weshaggard) [inline src/System.Runtime.InteropServices/tests/Configurations.props]: OK sounds good.

**Comment by weshaggard** [comment]
@dotnet-bot test OSX x64 Debug Build

File: src/System.Runtime.InteropServices/System.Runtime.InteropServices.sln [hunk 3]
```diff
@@ -26,10 +26,10 @@ Global
 		Release|Any CPU = Release|Any CPU
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
-		{A824F4CD-935B-4496-A1B2-C3664936DA7B}.Debug|Any CPU.ActiveCfg = netcoreapp-Debug|Any CPU
-		{A824F4CD-935B-4496-A1B2-C3664936DA7B}.Debug|Any CPU.Build.0 = netcoreapp-Debug|Any CPU
-		{A824F4CD-935B-4496-A1B2-C3664936DA7B}.Release|Any CPU.ActiveCfg = netcoreapp-Release|Any CPU
-		{A824F4CD-935B-4496-A1B2-C3664936DA7B}.Release|Any CPU.Build.0 = netcoreapp-Release|Any CPU
+		{A824F4CD-935B-4496-A1B2-C3664936DA7B}.Debug|Any CPU.ActiveCfg = uapaot-Debug|Any CPU
```
Comment (weshaggard) [inline src/System.Runtime.InteropServices/System.Runtime.InteropServices.sln]: I suspect the changes to this file will get reset once you run UpdateVSConfiguration target.

Comment (hughbe) [inline src/System.Runtime.InteropServices/System.Runtime.InteropServices.sln]: Correct -I've udpated it

File: src/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.Tests.csproj (hunk 2 above)
Comment (hughbe) [inline src/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.Tests.csproj]: I pushed a change that adds the configurations you suggested. I think these tests didn't work with uap/uapaot

**Comment by hughbe** [comment]
Test Linux x64 Release Build
Test Windows x86 Release Build

**Comment by hughbe** [comment]
TestOSX x64 Debug Build

**Comment by hughbe** [comment]
Test OSX x64 Debug Build

**Pull request merged**